server/src/main/java/io/druid/client/indexing/ClientCompactQuery.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.client.indexing;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.druid.timeline.DataSegment;
+
+import java.util.List;
+import java.util.Map;
+
+public class ClientCompactQuery
```

Comment:
hm, I feel like is more like a "Specification" than a "Query", maybe rename this to ClientCompactSpec

Reply:
The name of the taskSpecs used by overlords have a pattern of ```Client***Query```. I guess it's because the tasks submitted by coordinators can be thought of as queries to overlords. I'm not sure this is a good convention because it makes me somewhat confused at first.

I think it's better to move all task classes to some module (maybe  ```druid-api```) which allows all other modules and user libraries to depend on. This will let us to share the same class in all places. If we decide to go for it, I would like to do in a follow-up PR. What do you think?

Reply:
@jihoonson If you mean the actual "CompactionTask" etc classes, I think probably moving something so heavy from druid-indexing-service all the way down to druid-api would probably require collapsing a ton of druid modules into one giant druid-core module. I guess we could do that but it seems like a big change. Do you think it's worth it?

Reply:
I am definitely open to an argument that collapsing some modules is a good idea because it solves problems like this, and because the multi-module design of Druid didn't accomplish what it originally set out to accomplish (making it easy for people to write extensions or clients without depending on a ton of code). It's still required for many extensions, and basically all clients, to at least depend on druid-processing if not druid-server too.

If this is an argument that you want to make, could you please propose a list of modules to collapse and some comments about why?

Reply:
> I think probably moving something so heavy from druid-indexing-service all the way down to druid-api would probably require collapsing a ton of druid modules into one giant druid-core module.

@gianm would you elaborate on this? I simply thought some classes used as the input or return types for user APIs (like ```CompactionTask``` or ```TaskStatus```) can be moved to ```druid-api``` rather than collapsing core modules to a single giant one.

Reply:
@jihoonson I meant that the the API classes like `CompactionTask` are very heavy with implementation, and likely have enough transitive class dependencies that it is impractical to move them to `druid-api` without pulling along a lot of other classes too. However, like I said I am open to an argument that this is worth doing, since I'm not sure the large number of core druid modules is buying us much vs. a small number of larger modules.

Reply:
maybe in the far future... we can probably start by bringing druid-api, druid-commons, druid-processing, java-utils in a druid-core module as pretty much all extensions depend on them anyway.
later we can possibly move some core stuff from druid-server and druid-indexing-service to core as well.

Reply:
also, I do agree with @jon-wei on naming it *spec instead of *query even if that does not match with existing not-so-great convention . we don't need to change others but this class could have the saner name . "ClientCompactQuery" gave me the perception that it was a new query type.

Reply:
nvrmnd, as I look at more code .. I think its very unintuitive and probably should be fixed in a separate PR. to me, there are two weird things about these classes.
- xxxQuery gives the impression that they are some query types as in groupBy
- xxxQuery serialized json is expected to be deserializable into a Task object. so, probably we should use corresponding Task classes themselves rather that creating these new xxxQuery (which are tasks really)

Reply:
https://github.com/druid-io/druid/issues/5222

Reply:
@himanshug thanks for raising #5222. I think the best solution is to use the same class in ```druid-server``` and ```druid-indexing-service``` modules. To do so, some classes should be moved to a common module which both ```druid-server``` and ```druid-indexing-service``` modules depend on. 

@gianm thanks for the explanation. Yes, there are a lot of class dependencies on indexing task implementations, so collapsing several druid modules might be better. I will raise an issue to discuss it.